Mlp 206 p202] create page for creating/editing a role

alcabelu opened the pull request:
# Background

Webapp required a page for creating a role, and page for delete a role by roleId.
this pages consuming a WCF host in http://localhost:10160/Service.svc/Jsonp/post with method post an in body json request 
for add 
{"requests":[{"__type":"RoleAddRequest:#P202.Training.WCF.RequestsAndResponses",
"Role":{"Name": "developer"}	
}]}

for delete

{"requests":[{"__type":"RoleDeleteRequest:#P202.Training.WCF.RequestsAndResponses",
"RoleId":1
}]}

# Changes Done

-	Add in routes,js for roleAdd, roleDelete
-	Add services in roleServices.js  for Add and Delete
-	Added RoleAdd.Html, RoleAdd.js
-	Added RoleDelete.Html, RoleDetele.js-	
-	Added in RoleList the button for delete and check for edit
-	Added datetime of insert in Dominio of WebServices WCF




# Pending to be done

Add functionality for delete from rolelist with button get the ID from grid

# Notes

For execution in cmd  type  npm install  and npm start,
In a web browser open //localhost:8080
And select from the list RoleAdd or RoleDelete


Appovers required
@martinkeimel
@gmsoft
@darkainsoul

alcabelu committed:
add RoleAdd Html and JS

alcabelu committed:
Add service for add roleAdd

add check for edit in list roles
add button for delete roles

alcabelu committed:
Add entity Rol, for display a row in a grid of roles

alcabelu committed:
generate a simple object for rolinsert,

eliminate the console log statments

alcabelu committed:
add route for roleDelete page in routes.js

add service for roleDelete in roleService.js
add roleDelete.html
add roleDelete.js
remove console log in roleadd.html

martinkeimel commented on webapp/app/services/roleService.js:
```diff
@@ -1,17 +1,43 @@
-      data: '{"requests":[{"__type":"RoleReadAllRequest:#P202.Training.WCF.RequestsAndResponses"}]}'
-    });
-  }
+        console.log('Entre a llamar WebService')
```
Minor: please remove `console.log` lines before merging

alcabelu replied:
remove done

martinkeimel commented on webapp/app/services/roleService.js:
```diff
@@ -1,17 +1,43 @@
+
+    function roleDelete(value) {
+        // TODO: refactor common headers and base url
+        console.log('roleServices.js/rokeDelete');
```
Minor: please remove console.log lines before merging

alcabelu replied:
i remove the console.log

martinkeimel commented on webapp/app/services/roleService.js:
```diff
@@ -1,17 +1,43 @@
-  return {
-      roleReadAll: roleReadAll
-  };
+    function roleAdd(value) {
```
this `value` its actually the `roleName` right? perhaps you can rename it so it's clearer

alcabelu replied:
rename parameters was done

martinkeimel commented on webapp/app/services/roleService.js:
```diff
@@ -1,17 +1,43 @@
+        var roleIns = {
+            "Name":value
+        };              
+        var roleInsJs = JSON.stringify(roleIns);
```
you can replace the previous 4 lines with sth like `var role = '{ "Name": ' + roleName + '}';`

alcabelu replied:
the lines was replace with only one line of code

martinkeimel commented on webapp/app/services/roleService.js:
```diff
@@ -1,17 +1,43 @@
+        });
+    }
+
+    function roleDelete(value) {
```
can you rename the `value` parameter with something more descriptive?

alcabelu replied:
off course i rename parameter to roleId

martinkeimel commented on webapp/app/viewModels/role/RoleAdd.js:
```diff
@@ -0,0 +1,22 @@
+
+    function roleAdd() {
+        roleService.roleAdd(this.name()).then(function (response) {            
+            vm.roleResponse(response.ProcessJsonRequestsPostResult[0].Value);
```
You should redirect the user to the role list here

martinkeimel commented on webapp/app/viewModels/role/RoleDelete.js:
```diff
@@ -0,0 +1,23 @@
+require('services/roleService');
```
I think you don't need a new page+vm in order to delete a role. Since you already have a role list you can just make the call to the API from a button in that page

martinkeimel commented on webapp/app/viewModels/role/roleDelete.html:
```diff
@@ -0,0 +1,15 @@
+<section id="roleAdd">
```
Same here: you don't really need to create a new page just for deleting

martinkeimel commented on webapp/app/viewModels/role/roleAdd.html:
```diff
@@ -0,0 +1,15 @@
+        </span>
+    </div>
+    <button type="button" data-bind="click: roleAdd">Add</button>
+    <div>
```
Since you'll be redirecting the user to the role list after calling the API perhaps you can remove the response section

martinkeimel commented on webapp/app/viewModels/role/roleList.js:
```diff
@@ -1,6 +1,17 @@
+
+
+    // Class to represent a row in the seat reservations grid
+    function rol(id,name,createdby) {
```
You're not using this function right? Perhaps you can remove it

martinkeimel reviewed (changes requested):
Hi Luigi! I left a few comments with some changes required before merging

alcabelu committed:
remove console.log

rename parameters for more descriptive name

alcabelu committed:
Add route to RoleList

alcabelu committed:
se adiciona el button de delete con parent.remove

en el html
se adiciona la fuincon remove dentro del vm para ejecutar el Delete del rol

alcabelu committed:
se eliminan los console.log

se realiza el llamado al servicio de delete con role.ID directo
elimnando la variable id

martinkeimel commented on P202-Training.Domain/RoleService.cs:
```diff
@@ -49,6 +50,7 @@ public IList<Role> ListRoles()
 
         public Role ReadRole(int id)
         {
+           
```
Unneeded new empty line

alcabelu replied:
i remove the empty line

alcabelu committed:
se eliminan los archivos roleDelete de html y js

in routes js remove route to routedelete

martinkeimel commented on webapp/app/viewModels/role/roleList.js:
```diff
@@ -1,20 +1,25 @@
+    roleResponse: ko.observable(''),
+    remove: function (role) {       
+        roleService.roleDelete(role.Id).then(function (response) {
+            vm.roleResponse(response.ProcessJsonRequestsPostResult[0].Value);
```
What do you think about showing a nice message instead of the raw response? Something like "The role {roleName}has been successfully removed"

martinkeimel commented on webapp/app/viewModels/role/roleList.html:
```diff
@@ -2,13 +2,19 @@
         </thead>
         <tbody data-bind="foreach : roles">
             <tr>
+                <td>
```
What do you need this checkbox for?

alcabelu replied:
i have a idea for mark a rol for edit in the grid, change label for text, but in the talk i have with you , the edition are in the page of create, i add a button for ediiting and remove the checkbox

martinkeimel commented on webapp/app/services/roleService.js:
```diff
@@ -1,17 +1,38 @@
-      roleReadAll: roleReadAll
-  };
+    function roleAdd(roleName) {
+        var role = {"Name":roleName};              
```
Do you need this auxiliary variable? perhaps you can solve the stringify without that

alcabelu replied:
i move the object role to stringify

martinkeimel commented on webapp/app/viewModels/role/RoleAdd.js:
```diff
@@ -0,0 +1,23 @@
+
+    function roleAdd() {
+        roleService.roleAdd(this.name()).then(function (response) {  
+            var url = '#roleList/';
```
Same here: Do you need this auxiliary variable?

alcabelu replied:
i remove the var url

alcabelu committed:
remove empty line in file Domain RoleServices.cs

remove var role for stringify in file roleservices.js
remove checkbox in table in file roleList.html
remove var url in file roleAdd.js

alcabelu committed:
Add a most friendly message for information about delete

alcabelu committed:
in the file roleAdd.html remove the div of response

in the file roleList.js add control of error if not delete

alcabelu committed:
Add control of catch for promise of call webservice

alcabelu committed:
se adiciona el vm.roles.remove
